app/wasm/bindings/types_test.go

```diff
@@ -0,0 +1,122 @@
+package wasm
+
+import (
+	"encoding/json"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+var (
+	swap = Swap{
+		PoolId:   1,
+		DenomIn:  "denom_in",
+		DenomOut: "denom_out",
+	}
+	step = Step{
+		PoolId:   2,
+		DenomOut: "denom_out",
+	}
+	in           = sdk.NewInt(123)
+	out          = sdk.NewInt(456)
+	swapAmountIn = SwapAmount{
+		In: &in,
+	}
+	swapAmountOut = SwapAmount{
+		Out: &out,
+	}
+	exactIn = ExactIn{
+		Input:     sdk.NewInt(789),
+		MinOutput: sdk.NewInt(101112),
+	}
+	exactOut = ExactOut{
+		MaxInput: sdk.NewInt(131415),
+		Output:   sdk.NewInt(161718),
+	}
+	swapAmountExactIn = SwapAmountWithLimit{
+		ExactIn: &exactIn,
+	}
+	swapAmountExactOut = SwapAmountWithLimit{
+		ExactOut: &exactOut,
+	}
+)
+
+func TestTypesEncodeDecode(t *testing.T) {
+	// Swap
+	// Marshal
+	bzSwap, err := json.Marshal(swap)
+	require.NoError(t, err)
+	// Unmarshal
+	var swap1 Swap
+	err = json.Unmarshal(bzSwap, &swap1)
+	require.NoError(t, err)
+	// Check
+	assert.Equal(t, swap, swap1)
+
+	// Step
+	// Marshal
+	bzStep, err := json.Marshal(step)
+	require.NoError(t, err)
+	// Unmarshal
+	var step1 Step
+	err = json.Unmarshal(bzStep, &step1)
+	require.NoError(t, err)
+	// Check
+	assert.Equal(t, step, step1)
+
+	// SwapAmount
+	// Marshal
+	bzSwapAmount, err := json.Marshal(swapAmountOut)
+	require.NoError(t, err)
+	// Unmarshal
+	var swapAmount1 SwapAmount
+	err = json.Unmarshal(bzSwapAmount, &swapAmount1)
+	require.NoError(t, err)
```

Comment:
This is not a sufficient check.

This just checks if the Go parser can parse the JSON. Not the Rust parser, which is notably a lot more demanding on the difference between `"denom": null` and `"denom": ""`. It would be good to assert the actual JSON string that was marshalled to.

Or parse into `map[string]interface{}` and assert there was exactly one key set (which is what we want... for enum, we cannot even have `"in": "foo", "out": null`)

Reply:
I understand. What I can do is provide a list of json strings, and then unmarshall them, instead of marshalling first and unmarshalling.